docs/design/fabric-v2+/fpc-registration.puml

```diff
@@ -12,261 +12,155 @@
-  end note
-  return ATTESTATION
-
-  |||
-
-  Manager1 -> Manager1 : mark ECC_ID as installed
-  note right
-    Manager1 maintains a list of
-    locally "installed/created" enclaves
+  group Persistent Credentials (to be discussed)
+  Enclave1 -> Enclave1 : seal state
+  note left
+    sealed state includes: ...
+    CC_Id, CC_Version,
+    MSPID, Channel_Id,
+    Enclave_SK, Enclave_DK, CC_DK,
+    TLCC_MR
   end note
-  return ATTESTATION, ECC_S_PK / error
-
-  deactivate Admin
-  |||
+  end group
+
+  return Credentials protobuf
+
+  return Credentials protobuf
+
+  Admin -> IAS ++: send quote
+  return Verification_Report
+  Admin -> Admin: insert Verification_Report in Credentials
+
+
+'  note over Admin1 #FFaa00
+'    **MST**: From an attestation generalization perspective i think there should be just a single
+'    call ""GetAttestationProof"" or alike rather than a separate place where you get quote
+'    and another where you get the registration report which requires some re-organization of calls.
+'    A few observations on this:
+'    - we do not strictly have to call the enclave via ECC. ECCManager could also directly call
+'      enclave.so. this would make this creation and key-management potentially easier?
+'    - the quote generation is not driven from within the enclave as currently shown above but from
+'      the outside!
+'    I think from a high-level there should be following three high-level calls either in
+'    ECC or ECCMAnager depending on above:
+'    - EnclaveCreate (key-gen & imprint of context, i.e., current create/binding flows
+'    - EnclaveProof: gets org endorsement-sig, creates report, then quote
+'      and, if epid, gets IAS report. 
+'    - EnclavePersist: returns sealed state (important for security that this is called only after
+'      EnclaveProof [which we might want to enforce using some state-machine?] )
+'  end note
+'  return ATTESTATION
 
 end
 
-group register attestation
-
-  activate Admin
-
-  Admin -> ERCC1 ++: register (CC_NAME, CC_VER, MSP_ID_ORG_1, SIG_Org_1, ECC_S_PK, ECC_E_PK, ATTESTATION)
-  group verify and store
-
-    ERCC1 -> ERCC1 : check attestation signature & status
-    note left
-      - IAS verification key defined in ERCC definition
-      - status denotes the attestation result
-    end note
-
-    ERCC1 -> ERCC1 : Retrieve ECC definition for <CC_NAME, CC_VER>
+group enclave registration
+  Admin -> ERCC1 ++: invoke register (Credentials)
+  Admin -> ERCC2 : invoke register (Credentials)
 
-    ERCC1 -> ERCC1 : Verify SIG_Org_1
+  ref over ERCC2
+    (same as ERCC1)
+  end ref
+  activate ERCC2
 
-    ERCC1 -> ERCC1 : check consistency of quote (report-body)
-    note left
-      ECC_MR (mrenclave of ecc) is defined in ECC definition;
-      TLCC_MR (mrenclave of tlcc) is defined in ERCC definition
-      CH_Id is read from ledger or defined in ERCC definition
-    end note
+  ERCC1 -> ERCC1 : check verification report (through hard-coded certificate)
```

Comment:
what seems missing here is the check of the consistency of the params in the registration and the chaincode definition. Ideally ercc still would be able to get that directly locally through some cctosyscall or alike.  The replication of the whole approval/commit handshake (and hence "outsourcing" the consistency check to a quorum of admins) could be a (last) alternative. But it definitely should be explicit as this is crucial for security.

Reply:
> what seems missing here is the check of the consistency of the params in the registration and the chaincode definition. Ideally ercc still would be able to get that directly locally through some cctosyscall or alike.

Good catch, I have added it.

> The replication of the whole approval/commit handshake (and hence "outsourcing" the consistency check to a quorum of admins) could be a (last) alternative. But it definitely should be explicit as this is crucial for security.

We can avoid this alternative on registration -- we'll revisit this for chaincode upgrades.